driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -280,21 +267,37 @@ public void doMaintenance() {
         }
     }
 
-    private PooledConnection getPooledConnection(final long timeout, final TimeUnit timeUnit) {
-        UsageTrackingInternalConnection internalConnection = pool.get(timeout, timeUnit);
-        while (shouldPrune(internalConnection)) {
-            pool.release(internalConnection, true);
-            internalConnection = pool.get(timeout, timeUnit);
-        }
-        if (LOGGER.isTraceEnabled()) {
-            LOGGER.trace(format("Checked out connection [%s] to server %s", getId(internalConnection), serverId.getAddress()));
+    private PooledConnection getPooledConnection(final Timeout timeout) throws MongoTimeoutException {
+        boolean externalTimeoutException = false;
+        try {
+            UsageTrackingInternalConnection internalConnection = pool.get(timeout.remainingOrInfinite(NANOSECONDS), NANOSECONDS);
+            while (shouldPrune(internalConnection)) {
+                pool.release(internalConnection, true);
+                long remainingNanos = timeout.remainingOrInfinite(NANOSECONDS);
+                if (Timeout.expired(remainingNanos)) {
+                    externalTimeoutException = true;
+                    throw createTimeoutException(timeout);
+                }
+                internalConnection = pool.get(remainingNanos, NANOSECONDS);
+            }
+            if (LOGGER.isTraceEnabled()) {
+                LOGGER.trace(format("Got connection [%s] to server %s", getId(internalConnection), serverId.getAddress()));
```

Comment:
The message was "Checked out", which was misleading. According to the [CMAP specification](https://github.com/mongodb/specifications/blob/master/source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst#checking-out-a-connection): _"A Connection MUST NOT be checked out until it is established"_. The fact that we got a connection from `DefaultConnectionPool.pool` does not mean that it had been opened.

Reply:
Should we just remove this log message if it's not a relevant state to be reported?  "Got" just seems non-specific enough as to be un-useful

Reply:
If it's not relevant now, it then it was not relevant even previously: getting a connection from a pool could require opening it, and opening could fail. So despite the log message claims that a connection was checked out, it might have been not the case. It appears to me that at least some of the logging in the `DefaultConnectionPool` should have been done via a logging `ConnectionPoolListener` implementation.

Done.